Add vector_extent and vector_simplify_extent

Algunenano opened the pull request:
WIP: Requires rebase after https://github.com/CartoDB/Windshaft/pull/646 is merged

Adds the capability to setup the layer/tile extent for MVTs by adding 2 new options in MapConfig: `vector_extent` and `vector_simplify_extent`.

Even though the `pg-mvt` can modify the extent per tile request, the `mapnik` renderer doesn't, and that's why it is set during the renderer creation; this means that changing it requires re-instantiating the map.

It is required that all layers of a map have the same values. The defaults are:
- 4096 / 256 if non declared.
- vector_extent / vector_extent if only vector_extent is declared.
- 4096 / vector_simplify_extent if only vector_simplify_extent is declared.

References CartoDB/Windshaft-cartodb#843

rochoa commented on lib/windshaft/renderers/mapnik/factory.js:
```diff
@@ -404,3 +408,20 @@ function prepareQuery(userSql, geomColumnName, geomColumnType, options) {
+    const extentsDefined = _.uniq(extents);
+
+    if (extentsDefined.length > 1) {
+        throw new Error("Multiple extent values in mapConfig (" + extents + ")");
```
We should document that we just support one extent for all the layers.

Algunenano replied:
It's in the doc:
https://github.com/CartoDB/Windshaft/blob/89b99cd0ce7cc17a20e197c46ec222a68ada1e49/doc/MapConfig-1.8.0.md#L181

Should I clarify it more?

rochoa replied:
I missed that. It's OK. :)

Algunenano commented on doc/MapConfig-1.8.0.md:
```diff
@@ -0,0 +1,359 @@
+    // OPTIONAL
+    // Version of this spec to use for validation.
+    // Defaults to "1.0.0".
+    version: "1.7.0",
```
:ok_hand:

Algunenano commented on doc/MapConfig-1.8.0.md:
```diff
@@ -0,0 +1,359 @@
+    // OPTIONAL
+    // Extent of the layer for MVTs. Must be the same for all layers
+    // Defaults to 4096
+    vector_layer_extent: 4096
```
I don't see `vector_extent` anywhere. I've used it's what is being used internally in `mapnik-vector-tile` but I don't mind changing it.

rochoa replied:
Hahaha. No, it doesn't exist anywhere. I meant that being an option within the layer, using `vector_layer_extent` looked like smurf naming.

Algunenano commented:
I've renamed `vector_layer_extent` into `layer_extent` and I've added an initial implementation for `vector_simplify_extent` with default to 256.
- For Mapnik it uses the configuration options to get the equivalent simplification.
- For St_AsMVT, for now, it uses an extra simplification step when needed (`ST_Simplify(ST_RemoveRepeatedPoints(${geomColumn}, ${tol}), ${tol}, true)` as geometry. This makes it compatible with Postgis 2.4, 2.5 and our current 2.4 fork. A function that does this as part of St_AsMVTGeom is likely to be added in the (near?) future to both our fork and Postgis 3.0 but how is still to be decided.

It is still pending adding some extra tests checking how both new MapConfig settings interact. I might need to increase tolerances around the MVT comparison functions since the result, specially when TWKB approximation is used, tend to differ more when `vector_extent` is much bigger than `vector_simplify_extent`; but it's probably not all that important.

Algunenano commented:
I've already been bitten by `vector_simplify_extent`'s default while doing some tests. I think it would be better to set it as `if vector_extent is defined, then use vector_extent, else 256`. This is compatible with the current buggy behaviour and would match my expectations when you only set `vector_extent`.

What do you think @rochoa ? Does it sound reasonable?

Algunenano commented:
This should be now ready for review. Keep in mind that sits on top of #646 (so it includes those changes too).

rochoa commented:
> What do you think @rochoa ? Does it sound reasonable?

I missed this. We talked off-line about it, though. I agree. 😃

rochoa commented on lib/windshaft/renderers/renderer_factory.js:
```diff
@@ -17,7 +17,7 @@ function RendererFactory(opts) {
     this.opts = opts;
 
-    this.mapnikRendererFactory = new MapnikRenderer.factory(opts.mapnik);
+    this.mapnikRendererFactory = new MapnikRenderer.factory(Object.assign(opts.mapnik, opts.mvt));
```
I guess this is not critical here but beware of the side-effects of `Object.assign`.

Algunenano replied:
This is no longer needed. I added it because I started adding the extent parameters as options, but since I ended up removing that I'll revert this change.

oleurud commented on lib/windshaft/renderers/pg-mvt/renderer.js:
```diff
@@ -45,13 +47,17 @@ module.exports = class PostgresVectorRenderer {
     getTile (z, x, y, callback) {
-        const emptyTileMsg = 'Tile does not exist';
-        const subqueries = this._getLayerQueries({ z, x, y });
+        const headers = {
```
:+1:

enekid commented on lib/windshaft/renderers/pg-mvt/factory.js:
```diff
@@ -139,15 +215,31 @@ PgMvtFactory.prototype = {
+
+                this.options.vector_simplify_extent = simplify_extent || DEFAULT_SIMPLIFY_EXTENT;
+
+                const columnNamesPromises = layers.map(layer => {
```
FYI: Just a little shorter notation when the arrow function consists of only a return statement:

```javascript
const columnNamesPromises = layers.map(layer => getLayerColumns(psql, layer););
```

enekid commented on lib/windshaft/renderers/pg-mvt/renderer.js:
```diff
@@ -25,18 +25,20 @@ function extractMVT(data) {
 ///
 module.exports = class PostgresVectorRenderer {
+
     constructor (layers, psql, attrs, options) {
```
default parameters:
```javascript
constructor (layers, psql, attrs, options = {}) {
```

Algunenano commented:
Superseeded by https://github.com/CartoDB/Windshaft/pull/659
